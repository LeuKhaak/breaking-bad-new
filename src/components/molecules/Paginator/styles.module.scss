@import "../../../assets/styles/styles.scss";

.numbers {
  display: flex;
  align-items: center;
  .pages {
    display: flex;
    align-items: center;
    margin-right: 2rem;
    .pageLabel {
      display: flex;
      align-items: flex-end;
      .pageInput {
        appearance: none;
      }
      .pageInput:checked ~ .pageNumber {
        @include pageButton;
        color: $gold;
        background: $white;
        box-shadow: 0 0.25rem 0.5rem $shadowDark;
        border-radius: 0.25rem;
      }
      .pageNumber {
        @include pageButton;
      }
    }
    .buttonTab {
      @include button;
    }
  }
  .buttonPrev,
  .buttonNext {
    @include button;
  }
  .buttonPrev {
    background-image: url(../../../assets/icons/arrow.svg);
    margin-right: 1.5rem;
  }
  .buttonNext {
    background-image: url(../../../assets/icons/arrow.svg);
    transform: rotateY(180deg);
  }
}
