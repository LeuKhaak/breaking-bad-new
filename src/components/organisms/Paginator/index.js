import React from "react";
import { useSelector, useDispatch } from "react-redux";
import * as styles from "./styles.module.scss";
import Btn from "src/components/atoms/Btn";
import arrow from "../../../assets/icons/arrow.svg";

function Paginator() {
  const currentPage = useSelector((state) => state.paginator.currentPage);
  const perPage = useSelector((state) => state.paginator.perPage);

  //console.log(currentPage);
  const p = Math.ceil(62 / perPage);
  const pages = [];
  for (let i = 1; i <= p; i++) {
    pages.push(i);
  }

  const displayPageNumber = (elem) => {
    let disp = "";
    switch (true) {
      case elem === 1:
        disp = "flex";
        break;
      case elem === p:
        disp = "flex";
        break;
      case elem > 3 && elem < p:
        disp = "none";
        break;
      default:
        disp = "flex";
    }
    return disp;
  };

  const displayPageDots = (elem) => {
    let disp = "";
    switch (true) {
      case elem === 5:
        disp = "flex";
        break;
      default:
        disp = "none";
    }
    return disp;
  };

  return (
    <div className={styles.numbers}>
      <div className={styles.pages}>
        {pages &&
          pages.map((el) => (
            <div key={el} className={styles.pageLabel}>
              <Btn
                btnText={el}
                btnStyle={
                  el === currentPage ? "pageNumberChecked" : "pageNumber"
                }
                disp={displayPageNumber(el)}
              />
              <Btn
                btnText="..."
                btnStyle="pageNumber"
                disp={displayPageDots(el)}
              />
            </div>
          ))}
      </div>
      <Btn icon={arrow} btnStyle="pageNumber" />
      <Btn icon={arrow} btnStyle="buttonNext" />
    </div>
  );
}

export default Paginator;
