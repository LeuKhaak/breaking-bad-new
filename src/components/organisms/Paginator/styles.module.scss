@import "../../../assets/styles/styles.scss";

.numbers {
  display: flex;
  align-items: center;
  .pages {
    display: flex;
    align-items: center;
    margin-right: 2rem;
  }
}
