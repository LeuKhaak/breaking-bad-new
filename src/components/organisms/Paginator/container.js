import React from "react";
import { useSelector, useDispatch } from "react-redux";
import { selectPage, prevPage, nextPage } from "src/store/actions/paginator";
import Paginator from "./component";

function PaginatorContainer() {
  const dispatch = useDispatch();
  const currentPage = useSelector((state) => state.paginator.currentPage);
  const currentTriad = useSelector((state) => state.paginator.currentTriad);
  const perPage = useSelector((state) => state.paginator.perPage);

  const p = Math.ceil(62 / perPage);
  const pages = [];
  let t = 1;
  let d = false;
  for (let i = 1; i <= p; i++) {
    pages.push([i, t, d]);
    if (i % 3 === 0) t++;
    if (i === 1 || i % 3 === 0) d = true;
    else d = false;
  }

  const displayPageNumber = (elem) => {
    let disp = "";
    switch (true) {
      case elem[0] === 1:
        disp = "flex";
        break;
      case elem[0] === p:
        disp = "flex";
        break;
      case elem[1] === currentTriad:
        disp = "flex";
        break;
      case currentPage === p && (elem[0] === p - 1 || elem[0] === p - 2):
        disp = "flex";
        break;
      default:
        disp = "none";
    }
    return disp;
  };

  const displayPageDots = (elem) => {
    let disp = "";
    switch (true) {
      case (elem[1] === currentTriad + 1 || elem[1] === currentTriad - 1) &&
        elem[2] &&
        elem[0] !== p:
        disp = "flex";
        break;
      default:
        disp = "none";
    }
    return disp;
  };

  const select = (val1, val2) => {
    dispatch(selectPage(val1, val2));
  };

  const prev = (val1, val2) => {
    dispatch(prevPage(val1, val2));
  };
  const prevP = currentPage > 1 ? currentPage - 1 : currentPage;
  const prevTriad =
    currentPage > 3 && (currentPage - 1) % 3 === 0
      ? currentTriad - 1
      : currentTriad;

  const next = (val1, val2) => {
    dispatch(nextPage(val1, val2));
  };
  const nextP = currentPage < p ? currentPage + 1 : currentPage;
  const nextTriad =
    currentPage < p && currentPage % 3 === 0 ? currentTriad + 1 : currentTriad;

  return (
    <Paginator
      pages={pages}
      currentPage={currentPage}
      currentTriad={currentTriad}
      displayPageNumber={displayPageNumber}
      displayPageDots={displayPageDots}
      select={select}
      prev={prev}
      prevP={prevP}
      prevTriad={prevTriad}
      next={next}
      nextP={nextP}
      nextTriad={nextTriad}
    />
  );
}

export const container = PaginatorContainer;
