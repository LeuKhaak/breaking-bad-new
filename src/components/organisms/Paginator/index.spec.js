import React from "react";
import { shallow } from "enzyme";
import Paginator from "./component";

let component = null;
const setComponent = (props) => shallow(<Paginator {...props} />);
beforeEach(() => {
  component = setComponent();
});

describe("Paginator testing:", () => {
  test("is Paginator rendered", () => {
    component = setComponent({
      pages: [
        [1, 1, false],
        [2, 1, true],
        [3, 1, false],
        [4, 2, true],
        [5, 2, false],
      ],
      displayPageNumber: jest.fn(),
      displayPageDots: jest.fn(),
      currentPage: 1,
    });
    expect(component).toMatchSnapshot();
  });
});
