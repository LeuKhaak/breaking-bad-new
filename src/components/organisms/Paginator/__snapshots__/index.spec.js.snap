// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Paginator testing: is Paginator rendered 1`] = `
<div>
  <div>
    <div
      key="1"
    >
      <Btn
        arg={1}
        arg2={1}
        btnStyle="pageNumberChecked"
        btnText={1}
      />
      <Btn
        arg={1}
        arg2={1}
        btnStyle="pageNumber"
        btnText="..."
      />
    </div>
    <div
      key="2"
    >
      <Btn
        arg={2}
        arg2={1}
        btnStyle="pageNumber"
        btnText={2}
      />
      <Btn
        arg={2}
        arg2={1}
        btnStyle="pageNumber"
        btnText="..."
      />
    </div>
    <div
      key="3"
    >
      <Btn
        arg={3}
        arg2={1}
        btnStyle="pageNumber"
        btnText={3}
      />
      <Btn
        arg={3}
        arg2={1}
        btnStyle="pageNumber"
        btnText="..."
      />
    </div>
    <div
      key="4"
    >
      <Btn
        arg={4}
        arg2={2}
        btnStyle="pageNumber"
        btnText={4}
      />
      <Btn
        arg={4}
        arg2={2}
        btnStyle="pageNumber"
        btnText="..."
      />
    </div>
    <div
      key="5"
    >
      <Btn
        arg={5}
        arg2={2}
        btnStyle="pageNumber"
        btnText={5}
      />
      <Btn
        arg={5}
        arg2={2}
        btnStyle="pageNumber"
        btnText="..."
      />
    </div>
  </div>
  <Btn
    btnStyle="pageNumber"
    icon="test-file-stub"
  />
  <Btn
    btnStyle="buttonNext"
    icon="test-file-stub"
  />
</div>
`;
